pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -2017,6 +2027,25 @@ private void handleLastMessageIdFromCompactedLedger(PersistentTopic persistentTo
         });
     }
 
+    private int calculateTheLastBatchIndexInBatch(MessageMetadata metadata, ByteBuf payload) throws IOException {
+        int batchSize = metadata.getNumMessagesInBatch();
+        if (batchSize <= 1){
+            return -1;
+        }
+        SingleMessageMetadata singleMessageMetadata = new SingleMessageMetadata();
+        int lastBatchIndexInBatch = -1;
+        for (int i = 0; i < batchSize; i++){
+            ByteBuf singleMessagePayload =
+                    Commands.deSerializeSingleMessageInBatch(payload, singleMessageMetadata, i, batchSize);
```

Comment:
Perhaps reverse order is better?

Reply:
Ah, we can't extract batch messages in reverse order. Same as https://github.com/apache/pulsar/pull/18877#discussion_r1045275815